### PR TITLE
Fix Brave Ads conversions

### LINKED ISSUE #4914: Fix Brave Ads conversions
### Steps to Reproduce 
  1. See https://github.com/brave/brave-browser/issues/19943

**Actual result:** 
Ads are not converted

**Expected result:**
Ads should be converted

**Reproduces how often:** [Easily reproduced, Intermittent Issue]
Easily reproduced

**Brave Version:** 

**Device details:** 

**Website problems only:**
- did you check with Brave Shields down?
- did you check in Safari/Firefox (WkWebView-based browsers)?

### PR DESCRIPTION
**Pull request by tmancey**
## Summary of Changes

This pull request fixes #4914

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`

## Test Plan:

See https://github.com/brave/brave-browser/issues/19943

## Reviewer Checklist:

- [x] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [x] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [x] Adequate unit test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable).
- [x] Issue and pull request is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: Client/Frontend/Browser/BrowserViewController/BVC+Rewards.swift [hunk 1]
```diff
@@ -292,28 +292,26 @@ extension Tab {
         
         let group = DispatchGroup()
         group.enter()
-        
-        webView.evaluateSafeJavaScript(functionName: "document.documentElement.outerHTML.toString", contentWorld: .defaultClient) { html, _ in
+
+        webView.evaluateSafeJavaScript(functionName: "new XMLSerializer().serializeToString(document)", contentWorld: .defaultClient) { html, _ in
```
Comment (tmancey) [inline Client/Frontend/Browser/BrowserViewController/BVC+Rewards.swift]: Updated to match desktop

File: Client/Frontend/Browser/BrowserViewController/BVC+Rewards.swift [hunk 2]
```diff
@@ -292,28 +292,26 @@ extension Tab {
...
         }
-        
+
         if shouldClassifyLoadsForAds {
             group.enter()
-            webView.evaluateSafeJavaScript(functionName: "document.body.innerText", contentWorld: .defaultClient, asFunction: false) { text, _ in
-                // Get the list of words in the page and join them together with a space
-                // to send to the classifier
-                classifierText = (text as? String)?.words.joined(separator: " ")
+            webView.evaluateSafeJavaScript(functionName: "document?.body?.innerText", contentWorld: .defaultClient, asFunction: false) { text, _ in
```
Comment (tmancey) [inline Client/Frontend/Browser/BrowserViewController/BVC+Rewards.swift:303]: Updated to match desktop

File: Client/Frontend/Browser/BrowserViewController/BVC+Rewards.swift [hunk 3]
```diff
@@ -292,28 +292,26 @@ extension Tab {
...
         }
-        
+
         group.notify(queue: .main) {
             let faviconURL = URL(string: self.displayFavicon?.url ?? "")
             if faviconURL == nil {
                 log.warning("No favicon found in \(self) to report to rewards panel")
             }
-            rewards.reportLoadedPage(url: url, redirectionURLs: urls,
+            rewards.reportLoadedPage(url: url, redirectionURLs: urls.isEmpty ? [url] : urls,
```
Comment (tmancey) [inline Client/Frontend/Browser/BrowserViewController/BVC+Rewards.swift:314]: Redirect chain on brave-core always has at least one entry even if there are no redirects

File: Client/Frontend/Browser/BrowserViewController/BVC+Rewards.swift (hunk 1 above)
Comment (Brandon-T) [reply Client/Frontend/Browser/BrowserViewController/BVC+Rewards.swift]: It should be: `webView.evaluateSafeJavascript(functionName: "new XMLSerializer().serializeToString", args: ["document"], escapeArgs: false)`

Your code would also work fine if you put `asFunction: false`.

File: Client/Frontend/Browser/BrowserViewController/BVC+Rewards.swift (hunk 2 above)
Comment (Brandon-T) [reply Client/Frontend/Browser/BrowserViewController/BVC+Rewards.swift:303]: Are you sure about this change? `document.body` shouldn't ever really be null, so not sure if you require the optional chaining operator (`?`) here.

**Review by Brandon-T — CHANGES_REQUESTED** [review]

**Comment by tmancey** [comment]
@jonathansampson spotted some issues on brave-core where the optional chaining was needed. May not affect desktop but keeps them in sync. Thanks

**Review by Brandon-T — APPROVED** [review]

**Review by iccub — APPROVED** [review]

**Pull request merged**